minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java

```diff
@@ -300,8 +336,14 @@ private void initializeCloudReceiver() {
             .setLocation(ipcIdentity.getLocation())
             .setSystemId(ipcIdentity.getId())
             .build();
-        asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
-        LOG.info("Initialized cloud receiver stream");
+
+        cloudToMinionStreamCancellableContext = Context.current().withCancellation();
+        cloudToMinionStreamCancellableContext.run(
+            () -> {
+                asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
```

Comment:
I think code needs to be wrapped into condition cause context can be cancelled before runnable is executed.

```suggestion
                if (!Context.current().isCancelled()) {
                    asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
                }
```

Reply:
Do you have more information?  I found this snippet recommended online and followed that pattern.

Reply:
Looking at that carefullly, what happens if we don't check `Context.current().isCancelled()` first?  Also, is there something we should be doing when it is cancelled?

Reply:
Oh it seems there's another concern that creates; from the javadocs:

`Create a new context which is independently cancellable and also cascades cancellation from its parent. Callers must ensure that either [Context.CancellableContext.cancel(Throwable)](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.CancellableContext.html#cancel-java.lang.Throwable-) or [Context.CancellableContext.detachAndCancel(Context, Throwable)](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.CancellableContext.html#detachAndCancel-io.grpc.Context-java.lang.Throwable-) are called at a later point, in order to allow this context to be garbage collected.`

Reply:
It is indeed note from javadoc and example of how to use cancellable context.  I suppose that calling `cancel(null)` when we shutdown channel or entire client is fine.

Reply:
Good news - we do call cancel on it normally, so we should not have a leak.  The `handleDisconnect` method ensures that.